// socketServer/socketServer.js
var SocketError = require('socketServer/error').SocketError;
var socketServer = require('socketServer');
var server = socketServer();
exports.init =  function(httpServer){
    server.start(httpServer);
}
exports.server = server;
exports.channel = server.channel;

//add routing for client request
require('routes/websocket')(server);

// default route
server.use( function(req, res, next){
    res.send({
        error: {
            message: "undefined request",
            status: 404
        }
    })
});

server.use(function(err, req, res, next){

    if( typeof err == "number"){
        err = new SocketError(err);
    }

    if( err instanceof SocketError ){
        res.send({
            error: {
                message: err.message,
                status: err.status
            }
        })
    }else if( err instanceof Error){

        res.send({
            error: {
                message: err.message,
                status: 500
            }
        })

    }else{
        //process.env.NODE_ENV == "development"

        res.send({
            error: {
                message: err,
                status: 500
            }
        });
    }
})
